src/internal.c

```diff
@@ -12679,6 +12951,222 @@ static int BuildUserAuthRequestPassword(WOLFSSH* ssh,
+            begin += authData->sf.keyboard.promptLengths[entry];
+            output[begin] = authData->sf.keyboard.promptEcho[entry];
+            begin++;
+        }
+        *idx = begin;
+    }
+
+    return ret;
+}
+
+int SendUserAuthKeyboardRequest(WOLFSSH* ssh, WS_UserAuthData* authData)
+{
+    byte* output;
+    word32 idx;
+    word32 payloadSz = 0;
+    int ret = WS_SUCCESS;
+
+    WLOG(WS_LOG_DEBUG, "Entering SendUserAuthKeyboardRequest()");
+
+
+    if (ssh == NULL || authData == NULL) {
+        ret = WS_BAD_ARGUMENT;
+    }
+
+    if (ret == WS_SUCCESS) {
+        ret = ssh->ctx->keyboardAuthCb(&authData->sf.keyboard,
+                                       ssh->keyboardAuthCtx);
+    }
+
+    if (authData->sf.keyboard.promptCount > 0 &&
+        (authData->sf.keyboard.prompts == NULL ||
+         authData->sf.keyboard.promptLengths == NULL ||
+         authData->sf.keyboard.promptEcho == NULL)) {
+
+        ret = WS_BAD_USAGE;
+    }
+
+    if (authData->sf.keyboard.promptCount > WOLFSSH_MAX_PROMPTS) {
+        ret = WS_BAD_USAGE;
+    }
+
+    ssh->kbAuth.promptCount = authData->sf.keyboard.promptCount;
+
+    payloadSz = MSG_ID_SZ;
+    if (ret == WS_SUCCESS) {
+        ret = PrepareUserAuthRequestKeyboard(ssh, &payloadSz, authData);
+    }
+
+    if (ret == WS_SUCCESS) {
+        ret = PreparePacket(ssh, payloadSz);
+    }
+
+    output = ssh->outputBuffer.buffer;
+    idx = ssh->outputBuffer.length;
+
+    output[idx++] = MSGID_USERAUTH_INFO_REQUEST;
+
+    if (ret == WS_SUCCESS) {
+        ret = BuildUserAuthRequestKeyboard(ssh, output, &idx, authData);
+    }
+
+    if (ret == WS_SUCCESS) {
+        ssh->outputBuffer.length = idx;
+        ret = BundlePacket(ssh);
+    }
+
+    if (ret == WS_SUCCESS) {
+        ret = wolfSSH_SendPacket(ssh);
+    }
+
+    if ((ret != WS_WANT_WRITE) && (ret != WS_SUCCESS)) {
+        PurgePacket(ssh);
+    }
+
+    WLOG(WS_LOG_DEBUG, "Leaving SendUserAuthKeyboardRequest(), ret = %d", ret);
+
+    return ret;
+}
+
+static int PrepareUserAuthResponseKeyboard(WOLFSSH* ssh, word32* payloadSz,
+        const WS_UserAuthData* authData)
+{
+    int ret = WS_SUCCESS;
+    word32 entry;
+
+    if (ssh == NULL || payloadSz == NULL || authData == NULL)
+        ret = WS_BAD_ARGUMENT;
+
+    if (ret == WS_SUCCESS)
+    {
+        *payloadSz += LENGTH_SZ;
+        for (entry = 0; entry < authData->sf.keyboard.responseCount; entry++) {
+            *payloadSz += LENGTH_SZ +
+                authData->sf.keyboard.responseLengths[entry];
+        }
+    }
+
+    return ret;
+}
+
+
+static int BuildUserAuthResponseKeyboard(WOLFSSH* ssh, byte* output, word32* idx,
+        const WS_UserAuthData* authData)
+{
+    int ret = WS_SUCCESS;
+    word32 begin;
+    word32 entry;
+
+    if (ssh == NULL || output == NULL || idx == NULL || authData == NULL)
+        ret = WS_BAD_ARGUMENT;
+
+    if (authData->sf.keyboard.promptCount != authData->sf.keyboard.responseCount) {
```

Comment:
If `authData` is NULL, `ret` is set to the error code, but we'll still dereference it here.

Reply:
Good catch. I ran my work from #769 and found another one too. Both fixed.